First version on how to bring your custom kube-prometheus-stack (#206)
kyma-project/examples: prometheus/README.md

## Overview

The Kyma monitoring stack often brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be reset at the next upgrade cycle.

As an alternative, you can install the upstream chart with all customization options parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.

> **CAUTION:** This tutorial describes a basic setup that you should not use in production. Typically, a production setup needs further configuration, like optimizing the amount of data to scrape and the required resource footprint of the installation. To achieve qualities like high availability, scalability, or durable long-term storage, you might need an even more advanced setup.

## Prerequisites

- Kyma as the target deployment environment.
- Kubectl > 1.22.x
- Helm 3.x

## Installation

### Preparation

1. Export your Namespace as a variable. Replace the `{namespace}` placeholder in the following command and run it:

    ```bash
    export KYMA_EXAMPLE_NS="{namespace}"
    ```

>     >Note: Please assure that this namespace has no Istio sidecar injection enabled. The helm chart will deploy jobs which will not succeed having sidecar injection enabled by default.

Reviewer: ```suggestion
    >Note: This Namespace must have **no** Istio sidecar injection enabled. The Helm chart deploys jobs that will not succeed when sidecar injection is enabled by default.
```

Isn't this actually a prerequisite? 
Wouldn't it be better as a fourth item under Prerequisites, something like that:

`- The Namespace you're working with must have Istio sidecar injection **disabled**.`

Author: You usually enable injection, so having it disabled is the default and nothing explicit.
Not sure about prerequisite.. then we should say you need a namespace mainly. As this point here talks about the namespace setup I think it is better to have it here,

    ```bash
    export HELM_RELEASE_NAME="{release-name}"
    ```

3. Update your Helm installation with the required Helm repository:

    ```bash
    helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
    helm repo update
    ```

### Install the kube-prometheus-stack

1. Run the Helm upgrade command, which installs the chart if it's not present yet. At the end of the command, change the Grafana admin password to some value of your choice.
    ```bash
    helm upgrade --install -n ${KYMA_EXAMPLE_NS} ${HELM_RELEASE_NAME} prometheus-community/kube-prometheus-stack -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/values.yaml --set grafana.adminPassword=myPwd
    ```

You can use the [values.yaml](./values.yaml) provided with this tutorial, which contains customized settings deviating from the default settings, or create your own one.
The provided `values.yaml` covers the following adjustments:
- Support parallel operation to a Kyma monitoring stack
- Support the scraping of workload secured with Istio strict mTLS

### Install Istio Support

> 

Reviewer: Why are the instructions in this sections not numbered, but the steps in other sections are? I'd prefer more consistency.

Author: because no one remembers about such details when writing it :P

Reviewer: Muahahaha I do 😈

> To configure Prometheus for scraping of the istio specific metrics from any istio-proxy running in the cluster, deploy a PodMonitor which scrapes any pod having a port with name `.*-envoy-prom` exposed.

Reviewer: ```suggestion
1. To configure Prometheus for scraping of the Istio-specific metrics from any istio-proxy running in the cluster, deploy a PodMonitor, which scrapes any Pod that has a port with name `.*-envoy-prom` exposed.
```

After adding the numbering, you must adjust the indentation for the commands below each numbered step.

```bash
kubectl -n ${KYMA_EXAMPLE_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/podmonitor-istio-proxy.yaml
```

> Also deploy a ServiceMonitor definition for the central metrics of the `istiod` deployment.

Reviewer: ```suggestion
2. Deploy a ServiceMonitor definition for the central metrics of the `istiod` deployment:
```

```bash
kubectl -n ${KYMA_EXAMPLE_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/servicemonitor-istiod.yaml
```

> As Grafana is configured to load dashboards dynamically from ConfigMaps in the cluster, Istio specific dashboards can be applied as well. To get the latest versions you can follow the official [instructions](https://istio.io/latest/docs/ops/integrations/grafana/#option-1-quick-start) or you take the prepared ones by calling:

Reviewer: ```suggestion
3. Get the latest versions of the Istio-specific dashboards.
   Grafana is configured to load dashboards dynamically from ConfigMaps in the cluster, so Istio-specific dashboards can be applied as well. 
   Either follow the [Istio quick start instructions](https://istio.io/latest/docs/ops/integrations/grafana/#option-1-quick-start), or take the prepared ones with the following command:
```

Let's see whether formatting turns out alright...

```bash
kubectl -n ${KYMA_EXAMPLE_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/configmap-istio-grafana-dashboards.yaml
kubectl -n ${KYMA_EXAMPLE_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/configmap-istio-services-grafana-dashboards.yaml
```

> Please have in mind that this setup will collect all istio metrics on a pod level. As this can lead to cardinality issues and metrics are only needed on service level, it is recommended to use an improved setup based on federation as described in the [istio documentation](https://istio.io/latest/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring).

Reviewer: ```suggestion
   > **NOTE:** This setup collects all Istio metrics on a Pod level, which can lead to cardinality issues. Because  metrics are only needed on service level, it is recommended to use an improved setup based on federation as described in the [Istio documentation](https://istio.io/latest/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring).
```

Wait, why are we first instructing the users to do it that way, and afterwards tell them it's recommended to do it in another way? Is that because the recommendation is for a production setup (similar to the caution note in line 9)? If yes, then we should spell that out.

Author: hmm.. it is a more complex setup and I'm not sure if people always need it as it is mainly required for setups with many workloads.. I will improve you r message to include that info

### Verify the installation

1. You should see several Pods coming up in the Namespace, especially Prometheus and Alertmanager. Assure that all Pods have the "Running" state.
2. Browse the Prometheus dashboard and verify that all "Status->Targets" are healthy. The following command exposes the dashboard on `http://localhost:9090`:
   ```bash

>    kubectl -n ${KYMA_EXAMPLE_NS} port-forward svc/${HELM_RELEASE_NAME}-kube-prometheus-stack-prometheus 9090

Reviewer 2: The service name is in my test `${HELM_RELEASE_NAME}-kube-prometh-prometheus`

   ```
3. Browse the Grafana dashboard and verify that the dashboards are showing data. The user `admin` is pre-configured in the Helm chart; the password was provided in your `helm install` command. The following command exposes the dashboard on `http://localhost:3000`:
   ```bash
   kubectl -n ${KYMA_EXAMPLE_NS} port-forward svc/${HELM_RELEASE_NAME}-grafana 3000:80
   ```

### Deploy a custom workload and scrape it

1. Follow the tutorial [monitoring-custom-metrics](./../monitoring-custom-metrics/), but use the steps above to verify that the metrics are collected.

### Scrape workload via annotations

> Instead of defining a ServiceMonitor per workload for setting up custom metric scraping, you can use a simplified way based on annotations. The used [values.yaml](./values.yaml) defines an `additionalScrapeConfig` which will scrape all pods and services having these annotations:

Reviewer: ```suggestion
Instead of defining a ServiceMonitor per workload for setting up custom metric scraping, you can use a simplified way based on annotations. The used [values.yaml](./values.yaml) defines an `additionalScrapeConfig`, which  scrapes all Pods and services that have these annotations:
```

```yaml
prometheus.io/scrape: "true"   # mandatory to enable automatic scraping

> prometheus.io/scheme: https    # optional, default is http. Use "https" to scrape using istio client certificates. Will only work for services (not pods)

Reviewer 2: I'm afraid that many deployments that bring these annotations by default will not have the https scheme. This leads for instance for the [Kiali example](https://github.com/kyma-project/examples/tree/main/kiali) to a target that is `down`. I am wondering if it might be better to replace `prometheus.io` by something custom, so that the user makes a conscious decision when annotating anyting.

Author: Found a solution to change the default value when there is the typical istio label present

prometheus.io/port: "1234"     # optional, configure the port under which the metrics are exposed
prometheus.io/path: /myMetrics # optional, configure the path under which the metrics are exposed
```

> You can try it out by removing the ServiceMonitor of the example used above and instead providing the annotations to the Service manifest.

Reviewer: ```suggestion
You can try it out by removing the ServiceMonitor from the previius example and instead providing the annotations to the Service manifest.
```

### Cleanup

> Run the following command to remove the installation from the cluster:

Reviewer: ```suggestion
1. To remove the installation from the cluster, call Helm:
```

> 1. Remove the stack by calling Helm:

Reviewer: ```suggestion
```

# Install a custom kube-prometheus-stack in Kyma

## Overview

> The Kyma monitoring stack often brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be reset at the next upgrade cycle.

Reviewer: Why "often"? 🤔 

```suggestion
The Kyma monitoring stack brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be reset at the next upgrade cycle.
```

> As an alternative, you can install the upstream chart with all customization options parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.

Reviewer: I missed this "in" last time :D 

```suggestion
As an alternative, you can install the upstream chart with all customization options in parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.
```

> **CAUTION:**
- This tutorial describes a basic setup that you should not use in production. Typically, a production setup needs further configuration, like optimizing the amount of data to scrape and the required resource footprint of the installation. To achieve qualities like [high availability](https://prometheus.io/docs/introduction/faq/#can-prometheus-be-made-highly-available), [scalability](https://prometheus.io/docs/introduction/faq/#i-was-told-prometheus-doesnt-scale), or [durable long-term storage](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), you need a more advanced setup.
- This example uses the latest Grafana version, which is under AGPL-3.0 and might not be free of charge for commercial usage.

## Prerequisites

- Kyma as the target deployment environment.
- Kubectl > 1.22.x
- Helm 3.x

## Installation

### Preparation

> 1. Assure that the Kyma monitoring stack running in your cluster is limited to detection of kubernetes resources in the kyma-system namespace only, so that there can be no side-effects with the additional custom stack:

Reviewer: ```suggestion
1. If you cluster was installed manually using the Kyma CLI, you must assure that the Kyma monitoring stack running in your cluster is limited to detection of Kubernetes resources only in the `kyma-system` Namespace. To rule out that there are any side effects with the additional custom stack, run:
```

> > **NOTE**: That step is only needed for clusters installed manually via the Kyma CLI

Reviewer: ```suggestion
```

    ```bash
    kyma deploy --component monitoring --value monitoring.prometheusOperator.namespaces.releaseNamespace=true
    ```

1. Export your Namespace as a variable. Replace the `{namespace}` placeholder in the following command and run it:

    ```bash
    export KYMA_NS="{namespace}"
    ```
1. If you haven't created the Namespace yet, now is the time to do so:
    ```bash
    kubectl create namespace $KYMA_NS
    ```

> >**Note**: This Namespace must have **no** Istio sidecar injection enabled so no `istio-injection` label present on the namespace. The Helm chart deploys jobs that will not succeed when sidecar injection is enabled by default.

Reviewer: ```suggestion
   >**Note**: This Namespace must have **no** Istio sidecar injection enabled; that is, there must be no `istio-injection` label present on the Namespace. The Helm chart deploys jobs that will not succeed when Istop sidecar injection is enabled.
```

- adding 3 blanks for proper indentation below the step
- removing "by default", because I assume the jobs will also fail if the injection was enabled manually.

1. Export the Helm release name that you want to use. It can be any name, but be aware that all resources in the cluster will be prefixed with that name. Replace the `{release-name}` placeholder in the following command and run it:
   ```bash
    export HELM_RELEASE="{release-name}"
    ```

1. Update your Helm installation with the required Helm repository:

    ```bash
    helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
    helm repo update
    ```

### Install the kube-prometheus-stack

1. Run the Helm upgrade command, which installs the chart if it's not present yet. At the end of the command, change the Grafana admin password to some value of your choice.
    ```bash

>     helm upgrade --install -n ${KYMA_NS} ${HELM_RELEASE} prometheus-community/kube-prometheus-stack -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/values.yaml --set grafana.adminPassword=myPwd

Reviewer 3: Default kyma setup does not have a `--denyNamespaces` setup. see [here](https://github.com/kyma-project/kyma/blob/5ea7dbd64a1bf4b21b7198983cfe5c08599919b8/resources/monitoring/values.yaml#L1745). So the monitoring operator coming from the Kyma would interfere with the operator installed by above command and would cause constant restarts of the various pods.

Author: fixed by one more prep step

    ```

> You can use the [values.yaml](./values.yaml) provided with this tutorial, which contains customized settings deviating from the default settings, or create your own one.

Reviewer: Isn't this also an instruction?

```suggestion
2. You can use the [values.yaml](./values.yaml) provided with this tutorial, which contains customized settings deviating from the default settings, or create your own one.
```

    ```bash
    kubectl -n ${KYMA_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/configmap-istio-grafana-dashboards.yaml
    kubectl -n ${KYMA_NS} apply -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/istio/configmap-istio-services-grafana-dashboards.yaml
    ```

   > **NOTE:** This setup collects all Istio metrics on a Pod level, which can lead to cardinality issues. Because  metrics are only needed on service level, for setups having a bigger amount of workloads deployed, it is recommended to use a setup based on federation as described in the [Istio documentation](https://istio.io/latest/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring).

### Verify the installation

1. You should see several Pods coming up in the Namespace, especially Prometheus and Alertmanager. Assure that all Pods have the "Running" state.
2. Browse the Prometheus dashboard and verify that all "Status->Targets" are healthy. The following command exposes the dashboard on `http://localhost:9090`:
   ```bash

>    kubectl -n ${KYMA_NS} port-forward svc/${HELM_RELEASE}-kube-prometh-prometheus 9090

Reviewer 3: ```suggestion
   kubectl -n ${KYMA_NS} port-forward svc/${HELM_RELEASE}-kube-prometheus-sta-prometheus 90900
```

   ```

> 3. Browse the Grafana dashboard and verify that the dashboards are showing data. The user `admin` is pre-configured in the Helm chart; the password was provided in your `helm install` command. The following command exposes the dashboard on `http://localhost:3000`:

Reviewer: ```suggestion
3. Browse the Grafana dashboard and verify that the dashboards are showing data. The user `admin` is preconfigured in the Helm chart; the password was provided in your `helm install` command. The following command exposes the dashboard on `http://localhost:3000`:
```

[preconfigured](https://www.merriam-webster.com/dictionary/pre-configured) is the standard spelling

   ```bash
   kubectl -n ${KYMA_NS} port-forward svc/${HELM_RELEASE}-grafana 3000:80
   ```

### Deploy a custom workload and scrape it

> 1. Follow the tutorial [monitoring-custom-metrics](./../monitoring-custom-metrics/), but use the steps above to verify that the metrics are collected.

Reviewer: ```suggestion
Follow the tutorial [monitoring-custom-metrics](./../monitoring-custom-metrics/), but use the steps above to verify that the metrics are collected.
```

Only 1 step so we don't need numbering ;)

### Scrape workload via annotations

> Instead of defining a ServiceMonitor per workload for setting up custom metric scraping, you can use a simplified way based on annotations. The used [values.yaml](./values.yaml) defines an `additionalScrapeConfig`, which  scrapes all Pods and services that have these annotations:

Reviewer: ```suggestion
Instead of defining a ServiceMonitor per workload for setting up custom metric scraping, you can use a simplified way based on annotations. The used [values.yaml](./values.yaml) defines an `additionalScrapeConfig`, which  scrapes all Pods and services that have the following annotations:
```

```yaml
prometheus.io/scrape: "true"   # mandatory to enable automatic scraping
prometheus.io/scheme: https    # optional, default is "http" if no Istio sidecar is used. When using a sidecar (Pod has label `security.istio.io/tlsMode=istio`), the default is "https". Use "https" to scrape workloads using Istio client certificates. Will only work for annotated services (not Pods)
prometheus.io/port: "1234"     # optional, configure the port under which the metrics are exposed
prometheus.io/path: /myMetrics # optional, configure the path under which the metrics are exposed
```

You can try it out by removing the ServiceMonitor from the previous example and instead providing the annotations to the Service manifest.

### Cleanup

> 1. To remove the installation from the cluster, call Helm:

Reviewer: ```suggestion
To remove the installation from the cluster, call Helm:
```

Only 1 step so we don't need numbering ;)

>     ```bash

Reviewer: ```suggestion
```bash
```

>     helm delete -n ${KYMA_NS} ${HELM_RELEASE}

Reviewer: ```suggestion
helm delete -n ${KYMA_NS} ${HELM_RELEASE}
```

>     ```

Reviewer: ```suggestion
```
```

# Install a custom kube-prometheus-stack in Kyma

## Overview

> The Kyma monitoring stack often brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be resetted at next upgrade cycle.

Reviewer: ```suggestion
The Kyma monitoring stack often brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be reset at the next upgrade cycle.
```

> An alternative can be a parallel installation of the upstream chart offering all customization options. This tutorial outlines how to achieve such installation in co-existent to the Kyma monitoring stack.

Reviewer: ```suggestion
As an alternative, you can install the upstream chart with all customization options parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.
```

## Prerequisites

- Kyma as the target deployment environment.
- kubectl > 1.22.x
- helm 3.x

## Installation

### Preparation

1. Export your Namespace as a variable. Replace the `{namespace}` placeholder in the following command and run it:

    ```bash
    export KYMA_EXAMPLE_NS="{namespace}"
    ```

> 1. Export the Helm release name which you want to use. It can be any name,, be aware that all resources in the cluster will be prefixed with that name. Replace the `{release-name}` placeholder in the following command and run it:

Reviewer: ```suggestion
2. Export the Helm release name that you want to use. It can be any name, but be aware that all resources in the cluster will be prefixed with that name. Replace the `{release-name}` placeholder in the following command and run it:
```

    ```bash
    export HELM_RELEASE_NAME="{release-name}"
    ```

> 2. Update your helm installation with the required helm repository:

Reviewer: ```suggestion
3. Update your Helm installation with the required Helm repository:
```

    ```bash
    helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
    helm repo update
    ```

### Install the kube-prometheus-stack

> 1. Run the Helm upgrade command which will install the chart if not present yet. Change the grafana admin password (at the end of the command) to some value of your choice

Reviewer: ```suggestion
1. Run the Helm upgrade command, which installs the chart if it's not present yet. At the end of the command, change the Grafana admin password to some value of your choice.
```

    ```bash
    helm upgrade --install -n ${KYMA_EXAMPLE_NS} ${HELM_RELEASE_NAME} prometheus-community/kube-prometheus-stack -f https://raw.githubusercontent.com/kyma-project/examples/main/prometheus/values.yaml --set grafana.adminPassword=myPwd
    ```

> Hereby, use the [values.yaml](./values.yaml) provided with this tutorial which contains customized settings deviating from the default settings, or create your own one.

Reviewer: ```suggestion
You can use the [values.yaml](./values.yaml) provided with this tutorial, which contains customized settings deviating from the default settings, or create your own one.
```

> The provided values.yaml covers the following adjustments:

Reviewer: ```suggestion
The provided `values.yaml` covers the following adjustments:
```

- Support parallel operation to a Kyma monitoring stack

> - Scraping of Istio strict mTLS workloads

Reviewer: ```suggestion
- Scrape Istio strict mTLS workloads
```
previous line has no gerund either

### Verify the installation

> 1. You should see several pods coming up in the namespace, especially prometheus and alertmanager. Assure that all pods are ending in a "Running" state.

Reviewer: ```suggestion
1. You should see several Pods coming up in the Namespace, especially Prometheus and Alertmanager. Assure that all Pods have the "Running" state.
```

> 2. Browse the prometheus dashboard and verify that all "Status->Targets" are healthy. Following command will expose the dashboard on `http://localhost:9090`

Reviewer: ```suggestion
2. Browse the Prometheus dashboard and verify that all "Status->Targets" are healthy. The following command exposes the dashboard on `http://localhost:9090`:
```

   ```bash
   kubectl -n ${KYMA_EXAMPLE_NS} port-forward svc/${HELM_RELEASE_NAME}-kube-prometheus-stack-prometheus 9090
   ```

> 2. Browse the grafana dashboard and verify that the dashboards are showing data. The user `admin` is pre-configured in the helm chart, the password was provided in your helm install command. Following command will expose the dashboard on `http://localhost:3000`:

Reviewer: ```suggestion
3. Browse the Grafana dashboard and verify that the dashboards are showing data. The user `admin` is pre-configured in the Helm chart; the password was provided in your `helm install` command. The following command exposes the dashboard on `http://localhost:3000`:
```

   ```bash
   kubectl -n ${KYMA_EXAMPLE_NS} port-forward svc/${HELM_RELEASE_NAME}-grafana 3000:80
   ```

### Deploy a custom workload and scrape it

> 1. Follow the tutorial [monitoring-custom-metrics](./../monitoring-custom-metrics/) but use the steps above for verifying that the metrics are collected.

Reviewer: ```suggestion
1. Follow the tutorial [monitoring-custom-metrics](./../monitoring-custom-metrics/), but use the steps above to verify that the metrics are collected.
```

### Cleanup

> Run the following commands to completely remove the example and all its resources from the cluster:

Reviewer: ```suggestion
Run the following commands to remove the example and all its resources from the cluster:
```

> 1. Remove the stack by calling helm:

Reviewer: ```suggestion
1. Remove the stack by calling Helm:
```

## Overview

The Kyma monitoring stack often brings limited configuration options in contrast to the upstream [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack) chart. Modifications might be reset at the next upgrade cycle.

As an alternative, you can install the upstream chart with all customization options parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.

> Be aware of that this tutorial describes a basic setup which is not to be used in production. Further configuration is usually required like optimizations in regards to the amount of data to scrape and the required resource footprint of the installation. Even a different setup might be required to achieve qualities like high availability, scalability or durable long-term storage.

Reviewer: Ooops, didn't notice this sentence last time. 

```suggestion
> **CAUTION:** This tutorial describes a basic setup that you should not use in production. Typically, a production setup needs further configuration, like optimizing the amount of data to scrape and the required resource footprint of the installation. To achieve qualities like high availability, scalability, or durable long-term storage, you might need an even more advanced setup.
```

As an alternative, you can install the upstream chart with all customization options parallel. This tutorial outlines how to set up such installation in co-existence to the Kyma monitoring stack.

> **CAUTION:** This tutorial describes a basic setup that you should not use in production. Typically, a production setup needs further configuration, like optimizing the amount of data to scrape and the required resource footprint of the installation. To achieve qualities like [high availability](https://prometheus.io/docs/introduction/faq/#can-prometheus-be-made-highly-available), [scalability](https://prometheus.io/docs/introduction/faq/#i-was-told-prometheus-doesnt-scale), or [durable long-term storage](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), you need a more advanced setup.

> > **CAUTION:** This example will use the latest Grafana version which is under AGPL-3.0 and might not be free of charge for commercial usage.

Reviewer: ```suggestion
> **CAUTION:** This example uses the latest Grafana version, which is under AGPL-3.0 and might not be free of charge for commercial usage.
```

1. Export your Namespace as a variable. Replace the `{namespace}` placeholder in the following command and run it:

    ```bash
    export KYMA_NS="{namespace}"
    ```

> 1. If you don't have it created yet, now is the time to do so:

Reviewer: ```suggestion
1. If you haven't created the Namespace yet, now is the time to do so:
```

    ```bash
    kubectl create namespace $KYMA_NS
    ```

> 1. Assure that your namespace has istio-injection disabled by having the proper label in place

Reviewer: ```suggestion
1. This Namespace must have **no** Istio sidecar injection enabled. The Helm chart deploys jobs that fail if sidecar injection is enabled. To disable Istio sidecar injection, set the label:
```

    ```bash
    kubectl label namespace $KYMA_NS istio-injection=disabled
    ```

>     >**Note**: This Namespace must have **no** Istio sidecar injection enabled. The Helm chart deploys jobs that will not succeed when sidecar injection is enabled by default.

Reviewer: ```suggestion
```

### Scrape workload via annotations

Instead of defining a ServiceMonitor per workload for setting up custom metric scraping, you can use a simplified way based on annotations. The used [values.yaml](./values.yaml) defines an `additionalScrapeConfig`, which  scrapes all Pods and services that have these annotations:

```yaml
prometheus.io/scrape: "true"   # mandatory to enable automatic scraping

> prometheus.io/scheme: https    # optional, default is http when no istio sidecar is used. When using a sidecar (pod has label security.istio.io/tlsMode=istio) the default will be "https". Use "https" to scrape workloads using istio client certificates. Will only work for annotated services (not pods)

Reviewer: ```suggestion
prometheus.io/scheme: https    # optional, default is "http" if no Istio sidecar is used. When using a sidecar (Pod has label `security.istio.io/tlsMode=istio`), the default is "https". Use "https" to scrape workloads using Istio client certificates. Will only work for annotated services (not Pods)
```
